File: day5/task2.py
from functools import cmp_to_key


with open("input.txt") as f:
    cmp = {}
    while (l := f.readline()) != "\n":
        n1, n2 = map(int, l.split("|"))
        cmp[(n1, n2)] = -1
        cmp[(n2, n1)] = 1

    updates = [list(map(int, l.split(","))) for l in f.readlines()]

key = cmp_to_key(lambda x, y: cmp[(x, y)])
updates = [su for u in updates if u != (su := sorted(u, key=key))]

res = sum(u[len(u) // 2] for u in updates)
print(res)

File: day5/task2.cpp
#include <bits/stdc++.h>
using namespace std;

int main(void) {
    unordered_map<string, int> cmp;
    string line{"|"};

    ifstream file{"input.txt"};

    while (line.find('|') != string::npos) {
        cmp[line] = true;
        cmp[string{line.begin() + line.find('|') + 1, line.end()} + "|" +
            string{line.begin(), line.begin() + line.find('|')}] = false;
        file >> line;
    }

    vector<vector<string>> updates;
    while (file >> line) {
        stringstream ss(line);
        vector<string> v;
        while (ss.good()) {
            string substr;
            getline(ss, substr, ',');
            v.push_back(substr);
        }
        updates.push_back(v);
    }
    file.close();

    int res{0};
    for (auto u : updates) {
        vector<string> su = u;
        sort(su.begin(), su.end(),
             [&cmp](auto x, auto y) { return cmp[x + "|" + y]; });
        if (su != u)
            res += stoi(su[su.size() / 2]);
    }

    cout << res << '\n';

    return 0;
}
